sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala

```diff
@@ -452,6 +452,17 @@ class Analyzer(
 
     def apply(plan: LogicalPlan): LogicalPlan = plan resolveOperators {
       case i @ InsertIntoTable(u: UnresolvedRelation, parts, child, _, _) if child.resolved =>
+        // A partitioned relation's schema can be different from the input logicalPlan, since
+        // partition columns are all moved after data columns. We Project to adjust the ordering.
+        val input = if (parts.nonEmpty) {
+          val (inputPartCols, inputDataCols) = child.output.partition { attr =>
+            parts.contains(attr.name)
+          }
+          Project(inputDataCols ++ inputPartCols, child)
+        } else {
+          child
+        }
```

Comment:
Let me show an example, 
```SQL
INSERT OVERWRITE TABLE $table
partition (p1='a',p2)
SELECT 'blarr3' as p2, 'blarr1'
```

Without this change, the output is like
```
+------+---+------+
|    c1| p1|    p2|
+------+---+------+
|blarr3|  a|blarr1|
+------+---+------+
```
After this change, the output becomes
```
+------+---+------+
|    c1| p1|    p2|
+------+---+------+
|blarr1|  a|blarr3|
+------+---+------+
```
Based on the [Hive specification](https://cwiki.apache.org/confluence/display/Hive/DynamicPartitions), 

>>In INSERT ... SELECT ... queries, the dynamic partition columns must be specified last among the columns in the SELECT statement and in the same order in which they appear in the PARTITION() clause.

Now, the new behavior does not follow the Hive spec. I think we should not do this.